src/dataflow/src/source/postgres.rs

```diff
@@ -157,12 +168,34 @@ impl PostgresSourceReader {
                         let a: Datum = row.get(n).into();
                         a
                     }));
-                    snapshot_tx.insert(mz_row).await?;
+                    try_recoverable!(snapshot_tx.insert(mz_row.clone()).await);
+                    try_fatal!(buffer.write(&try_fatal!(bincode::serialize(&mz_row))).await);
                 }
             }
         }
+        try_recoverable!(client.simple_query("COMMIT;").await);
+        Ok(())
+    }
 
-        client.simple_query("COMMIT;").await?;
+    /// Reverts a failed snapshot by deleting any processed rows from the dataflow.
+    async fn revert_snapshot<R: Read + Seek>(
+        &self,
+        snapshot_tx: &mut SourceTransaction<'_>,
+        mut reader: R,
+    ) -> Result<(), anyhow::Error> {
+        let rows = tokio::task::block_in_place(|| -> Result<Vec<Row>, anyhow::Error> {
+            let len = reader.seek(SeekFrom::Current(0))?;
+            reader.seek(SeekFrom::Start(0))?;
+            let mut rows = Vec::new();
+            while reader.seek(SeekFrom::Current(0))? < len {
```

Comment:
Not a huge deal but this will be making a seek() system call for every row. An alternative is to use [`Read::take()`](https://doc.rust-lang.org/std/io/trait.Read.html#method.take) which has a `limit()` function on it that you can use to know when it's EOF

Reply:
Oh, interesting! So we need to call `deserialize_from` in a loop, do you have a suggestion for how I could use `Read::take()` to do that? (I tried a few things that didn't work out.)